packages/aws-cdk-lib/aws-codebuild/lib/project.ts

```diff
@@ -1334,13 +1334,19 @@ export class Project extends ProjectBase {
     const hasEnvironmentVars = Object.keys(vars).length > 0;
 
     const errors = this.buildImage.validate(env);
+
+    errors.push(...this.validateLambdaBuildImage(this.buildImage, props));
+
     if (errors.length > 0) {
       throw new Error('Invalid CodeBuild environment: ' + errors.join('\n'));
     }
 
-    const imagePullPrincipalType = this.buildImage.imagePullPrincipalType === ImagePullPrincipalType.CODEBUILD
-      ? ImagePullPrincipalType.CODEBUILD
-      : ImagePullPrincipalType.SERVICE_ROLE;
+    // For Lambda compute, specifying imagePullPrincipalType is not supported
+    const imagePullPrincipalType = this.isLambdaBuildImage(this.buildImage)
+      ? undefined
+      : this.buildImage.imagePullPrincipalType === ImagePullPrincipalType.CODEBUILD
+        ? ImagePullPrincipalType.CODEBUILD
+        : ImagePullPrincipalType.SERVICE_ROLE;
```

Comment:
```suggestion
    const imagePullPrincipalType = this.buildImage.imagePullPrincipalType === ImagePullPrincipalType.CODEBUILD
      ? ImagePullPrincipalType.CODEBUILD
      : ImagePullPrincipalType.SERVICE_ROLE;
```
Let's keep it as it was. See https://github.com/aws/aws-cdk/pull/27934#discussion_r1390885565.